app/src/org/commcare/android/framework/SessionAwareCommCareActivity.java

```diff
@@ -1,11 +1,30 @@
 package org.commcare.android.framework;
 
+import android.os.Bundle;
+
+import org.commcare.android.util.SessionUnavailableException;
+
 /**
  * Manage redirection to login screen when session expiration occurs.
  *
  * @author Phillip Mates (pmates@dimagi.com)
  */
 public abstract class SessionAwareCommCareActivity<R> extends CommCareActivity<R> {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        try {
+            onCreateAware(savedInstanceState);
+        } catch (SessionUnavailableException e) {
+            SessionActivityRegistration.redirectToLogin(this);
+        }
```

Comment:
We've gone a bit back and forth on it. While I agree that high level handling has a lot of downsides, I think that the semantic meaning of sessionunavailable basically always has a clear outcome in session aware activities, and I don't want us to fall into a trap of trying to sneak away from the current activity. In your example, for instance, the activity just finishes itself hoping that it'll end up back in the right place, but there's not really any guarantee you will. You could easily end up with a weird workflow where CommCare just exits, or the user ends up back on the home screen for reasons that are unclear to them.

90% of hte time "I wanted session data, but there wasn't a session" means "try to log back in", it seems totally reasonable to me for us to encode that in a way that doesn't need us to write custom logic each time.